source/opt/common_uniform_elim_pass.h

```diff
@@ -0,0 +1,214 @@
+  // been loaded in the same block for types whose loads cannot cross blocks.
+  bool CommonUniformLoadElimBlock(ir::Function* func);
+
+  // Eliminate duplicated extracts of same id. Extract may be moved to same
+  // block as the id definition. This is primarily intended for extracts
+  // from uniform loads. Most effective if preceded by
+  // CommonUniformLoadElimination().
+  bool CommonExtractElimination(ir::Function* func);
+
+  // Initialize the label2SSA map entry for a block. Insert phi instructions
+  // into block when necessary. All predecessor blocks must have been
+  // For function |func|, first change all constant index
+  // access chain loads into equivalent composite extracts. Then consolidate 
+  // identical uniform loads into one uniform load. Finally, consolidate
+  // identical uniform extracts into one uniform extract. This may require
+  // moving a load or extract to a point which dominates all uses.
+  // Return true if func is modified. 
+  //
+  // This pass requires the function to have structured control flow ie shader
+  // capability. It also requires logical addressing ie Addresses capability
+  // is not enabled. It also currently does not support any extensions.
+  bool EliminateCommonUniform(ir::Function* func);
+
+  // Initialize extensions whitelist
+  void InitExtensions();
+
+  // Return true if all extensions in this module are allowed by this pass.
+  bool AllExtensionsSupported() const;
+
+  // Return true if all uses of varId are only through supported reference
+  // operations ie. loads and store. Also cache in supported_ref_vars_;
+  inline bool IsDecorate(uint32_t op) const {
+    return (op == SpvOpDecorate || op == SpvOpDecorateId);
+  }
+
+  inline void FinalizeNextId(ir::Module* module) {
+    module->SetIdBound(next_id_);
+  }
+
+  inline uint32_t TakeNextId() {
+    return next_id_++;
+  }
+
+  void Initialize(ir::Module* module);
+  Pass::Status ProcessImpl();
+
+  // Module this pass is processing
+  ir::Module* module_;
+
+  // Def-Uses for the module we are processing
+  std::unique_ptr<analysis::DefUseManager> def_use_mgr_;
+
+  // Map from function's result id to function
+  std::unordered_map<uint32_t, ir::Function*> id2function_;
+
+  // Map from block's label id to block.
+  std::unordered_map<uint32_t, ir::BasicBlock*> id2block_;
+
+  // Map from block to its structured successor blocks. See 
+  // ComputeStructuredSuccessors() for definition.
+  std::unordered_map<const ir::BasicBlock*, std::vector<ir::BasicBlock*>>
+      block2structured_succs_;
+
+  // Map from block's label id to its predecessor blocks ids
+  std::unordered_map<uint32_t, std::vector<uint32_t>> label2preds_;
+
+  // Map from uniform variable id to its common load id
+  std::unordered_map<uint32_t, uint32_t> uniform2load_id_;
+
+  // Map of extract composite ids to map of indices to insts
+  std::unordered_map<uint32_t, std::unordered_map<uint32_t,
```

Comment:
Because coalescing involves moving the first extract (ie some work and potential downstream register pressure), I didn't want to initiate the move until I knew I had two or more extracts to coalesce. So I collected them first in this data structure and only do the coalesce when >=2 instructions are present.